### ISSUE TITLE
Could you help me? What's wrong with my dagnn?

### ISSUE THREAD

**Issue by hunterlew** [vlfeat/matconvnet#771]
I have tried 'dagnn' on my own database but got very large objectives and soon NAN. I could not find the reasons for that. Could you help me? The network initialization shown below:

function net = test_dagnn_init(varargin)

opts.batchNormalization = true;
opts.networkType = 'dagnn';
opts.cudnnWorkspaceLimit = 1024*1024*1024 ; % 1G
opts = vl_argparse(opts, varargin) ;

net = dagnn.DagNN() ;

lastAdded.var = 'input' ;
lastAdded.depth = 1 ; % image channel --byhunterlew

function Conv(name, ksize, depth, varargin)
% Helper function to add a Convolutional + BatchNorm + ReLU
% sequence to the network.
  args.relu = true ;
  args.downsample = false ;
  args.bias = true ;
  args = vl_argparse(args, varargin) ;
  if args.downsample, stride = 2 ; else stride = 1 ; end
  if args.bias, pars = {[name '_f'], [name '_b']} ; else pars = {[name '_f']} ; end
  net.addLayer([name  '_conv'], ...
               dagnn.Conv('size', [ksize ksize lastAdded.depth depth], ...
                          'stride', stride, ....
                          'pad', (ksize - 1) / 2, ...
                          'hasBias', args.bias, ...
                          'opts', {'cudnnworkspacelimit', opts.cudnnWorkspaceLimit}), ...
               lastAdded.var, ...
               [name '_conv'], ...
               pars) ;
  net.addLayer([name '_bn'], ...
               dagnn.BatchNorm('numChannels', depth, 'epsilon', 1e-5), ...
               [name '_conv'], ...
               [name '_bn'], ...
               {[name '_bn_w'], [name '_bn_b'], [name '_bn_m']}) ;
  lastAdded.depth = depth ;
  lastAdded.var = [name '_bn'] ;
  if args.relu
    net.addLayer([name '_relu'] , ...
                 dagnn.ReLU(), ...
                 lastAdded.var, ...
                 [name '_relu']) ;
    lastAdded.var = [name '_relu'] ;
  end
end

% -------------------------------------------------------------------------
% Add sections --byhunterlew
% -------------------------------------------------------------------------

Conv('conv_1',3,16);
net.addLayer('pool1', ...
             dagnn.Pooling('poolSize', [2 2]), ...
             lastAdded.var, ...
             'pool1') ;
lastAdded.var = 'pool1';

Conv('conv_2',3,32);
net.addLayer('pool2', ...
             dagnn.Pooling('poolSize', [2 2]), ...
             lastAdded.var, ...
             'pool2') ;
lastAdded.var = 'pool2';

Conv('conv_3',3,64);
net.addLayer('pool3', ...
             dagnn.Pooling('poolSize', [2 2]), ...
             lastAdded.var, ...
             'pool3') ;
lastAdded.var = 'pool3';

Conv('conv_4',3,128);
net.addLayer('pool4', ...
             dagnn.Pooling('poolSize', [2 2]), ...
             lastAdded.var, ...
             'pool4') ;
lastAdded.var = 'pool4';

net.addLayer('prediction_avg' , ...
             dagnn.Pooling('poolSize', [4 4], 'method', 'avg'), ...
             lastAdded.var, ...
             'prediction_avg') ;

net.addLayer('prediction' , ...
             dagnn.Conv('size', [1 1 128 10]), ...
             'prediction_avg', ...
             'prediction', ...
             {'prediction_f', 'prediction_b'}) ;

net.addLayer('loss', ...
             dagnn.Loss('loss', 'softmaxlog') ,...
             {'prediction', 'label'}, ...
             'objective') ;

net.addLayer('top1error', ...
             dagnn.Loss('loss', 'classerror'), ...
             {'prediction', 'label'}, ...
             'top1error') ;

net.addLayer('top5error', ...
             dagnn.Loss('loss', 'topkerror', 'opts', {'topK', 5}), ...
             {'prediction', 'label'}, ...
             'top5error') ;

% -------------------------------------------------------------------------
%                                                           Meta parameters
% -------------------------------------------------------------------------

net.meta.inputSize = [128 128 1] ;

lr = logspace(-2, -4, 100) ;
%lr = [0.1 * ones(1,30), 0.01*ones(1,30), 0.001*ones(1,30)] ;
net.meta.trainOpts.learningRate = lr ;
net.meta.trainOpts.numEpochs = numel(lr) ;
net.meta.trainOpts.momentum = 0.9 ;
net.meta.trainOpts.batchSize = 49 ;
net.meta.trainOpts.weightDecay = 0.0005 ;

% Init parameters randomly
net.initParams() ;

end

**Comment by albanie** [comment]
Hi @hunterlew - there a few generic steps which might be helpful for debugging exploding gradients (these are things I've found useful in the past):

1. Try with very low learning rates and check that the loss is actually going down (a good sanity check).
2. Try using a solver like [adaDelta](https://github.com/vlfeat/matconvnet/blob/new-solvers/examples/%2Bsolver/adadelta.m) (which sets its own learning rate) to see how learning is progressing over the first few epochs - this can be useful for getting an idea of what is going on.  
3. Try with an extremely simple network (minimal number of layers).
4. Check the norms of the weights are sensible.
